File: cpp/1034.coloring-a-border.cpp
/*
 * @lc app=leetcode id=1034 lang=cpp
 *
 * [1034] Coloring A Border
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
    int new_color;
    int og_color;
    int height;
    int width;
    vector<vector<int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

public:
    vector<vector<int>> colorBorder(vector<vector<int>> &grid, int row, int col, int color)
    {
        new_color = color;
        og_color = grid[row][col];

        height = grid.size();
        width = grid[0].size();
        helpler(grid, row, col);
        for (int i = 0; i != height; ++i)
        {
            for (int j = 0; j != width; ++j)
            {
                if (grid[i][j] < 0)
                {
                    grid[i][j] *= -1;
                }
            }
        }
        return grid;
    }
    bool helpler(vector<vector<int>> &grid, int row, int col)
    {
        if (grid[row][col] == -new_color || grid[row][col] == -og_color)
        {
            return true;
        }
        else if (grid[row][col] != og_color)
        {
            return false;
        }
        grid[row][col] = -new_color;
        bool is_inner = true;
        for (auto &&d : directions)
        {
            if (row + d[0] >= 0 && row + d[0] < height && col + d[1] >= 0 && col + d[1] < width)
            {
                if (!helpler(grid, row + d[0], col + d[1]))
                {
                    is_inner = false;
                }
            }
            else
            {
                is_inner = false;
            }
        }
        if (is_inner)
        {
            grid[row][col] = -og_color;
        }
        return true;
    }
};
// @lc code=end

File: py/1034.coloring-a-border.py
#
# @lc app=leetcode id=1034 lang=python3
#
# [1034] Coloring A Border
#

# @lc code=start
class Solution:
    def colorBorder(self, grid: List[List[int]], row: int, col: int, color: int) -> List[List[int]]:
        new_color = color
        og_color = grid[row][col]
        height = len(grid)
        width = len(grid[0])

        def helpler(row, col):
            if grid[row][col] == -new_color or grid[row][col] == -og_color:
                return True
            elif grid[row][col] != og_color:
                return False
            grid[row][col] = -new_color
            is_inner = True
            for r in range(-1, 2, 2):
                if row+r >= 0 and row+r < height:
                    if not helpler(row+r, col):
                        is_inner = False
                else:
                    is_inner = False
            for c in range(-1, 2, 2):
                if col+c >= 0 and col+c < width:
                    if not helpler(row, col+c):
                        is_inner = False
                else:
                    is_inner = False
            if is_inner:
                grid[row][col] = -og_color
            return True
        helpler(row, col)
        for i in range(height):
            for j in range(width):
                if grid[i][j] < 0:
                    grid[i][j] *= -1
        return grid

# @lc code=end
